measure/src/main/scala/org/apache/griffin/measure/step/builder/dsl/expr/TreeNode.scala

```diff
@@ -18,28 +18,42 @@ under the License.
 */
 package org.apache.griffin.measure.step.builder.dsl.expr
 
+import scala.reflect.ClassTag
+
 trait TreeNode extends Serializable {
 
   var children = Seq[TreeNode]()
 
   def addChild(expr: TreeNode) = { children :+= expr }
   def addChildren(exprs: Seq[TreeNode]) = { children ++= exprs }
 
-  def preOrderTraverseDepthFirst[T, A <: TreeNode](z: T)(seqOp: (A, T) => T, combOp: (T, T) => T): T = {
-    if (this.isInstanceOf[A]) {
-      val tv = seqOp(this.asInstanceOf[A], z)
-      children.foldLeft(combOp(z, tv)) { (ov, tn) =>
-        combOp(ov, tn.preOrderTraverseDepthFirst(z)(seqOp, combOp))
+  def preOrderTraverseDepthFirst[T, A <: TreeNode](z: T)(seqOp: (A, T) => T, combOp: (T, T) => T)(implicit tag: ClassTag[A]): T = {
+
+    val clazz = tag.runtimeClass
+
+    this.getClass match {
+      case `clazz` => {
```

Comment:
The code means this.getClass and T must be the same class. But in the usage, it traverses among different types of tree nodes, their concrete classes can't be matched to their parent class, which is not what we want.
I doubt the modification of this file will destroy the process, I suggest you write a unit test for this.

Reply:
we cannot use this patch, please remove this .

Reply:
@bhlx3lyx7 @grant-xuexu we will append a test unit for this.

Reply:
@bhlx3lyx7 yea, you are correct. I misunderstood the logic here. However the issue is `if(this.instanceOf[A])` could already return true because of the type erasure at runtime. 

There is a method called `isAssignableFrom` in Java Class[T], which could be used to check if a class is the subclass of another class. 

so the logic could be changed into:

`
val clazz = tag.runtimeClass

    if(clazz.isAssignableFrom(this.getClass)){
      val tv = seqOp(this.asInstanceOf[A], z)
      children.foldLeft(combOp(z, tv)) { (ov, tn) =>
        combOp(ov, tn.preOrderTraverseDepthFirst(z)(seqOp, combOp))
      }
    }
    else {
      z
    }
`